engine/swarm/secrets.md

```diff
@@ -0,0 +1,339 @@
+      DNS name `wordpress`.
+    - Because the scale is set to 1, only a single WordPress task runs.
+      Load-balancing WordPress is left as an exercise to the reader, because of
+      limitations with storing WordPress session data on the container filesystem.
+    - Exposes WordPress on port 30000 of the host machine, so that you can access
+      it from external hosts. You could expose port 80 if you do not have a web
+      server running on port 80 already.
+    - Grants the service access to the `mysql_password` secret, but specify a
+      different target file name within the container. The WordPress container
+      will use the mount point `/run/secrets/wp_db_password`. Also specify that
+      the secret is not world-readable, by setting the mode to `0400`.
+    - Sets the environment variable `WORDPRESS_DB_PASSWORD_FILE` to this file
+      path. The `wordpress` service will read the password string from that file.
+
+    ```bash
+    $ docker service create --name="wordpress" \
+                            --network="wp_network" \
+                            --replicas="1" \
+                            --publish 30000:80 \
+                            --secret source=mysql_password,target=wp_db_password,mode=0400 \
+                            -e WORDPRESS_DB_PASSWORD_FILE="/run/secrets/wp_db_password" \
+                            -e WORDPRESS_DB_HOST="mysql:3306" \
+                            wordpress:latest
+    ```
+
+7.  Access `http://localhost:30000/` from any Swarm node and begin using
+    WordPress.
+
+## Example: Rotate a secret
+
+This example builds upon the previous one. In this scenario, you create a new
+secret with the new MySQL password, update the `mysql` and `wordpress` services
+to use it, then remove the old secret.
+
+1.  Create the new secret, and name it `mysql_password_v2`.
+
+    ```bash
+    $ openssl rand -base64 20 |docker secret create mysql_password_v2
+      lwxgks2rojdbvaui5kay1suvr
+    ```
+
+2.  Update the `mysql` service to use the new password, keeping the target path
+    at `/run/secrets/mysql_password` within the container.
+
+    ```bash
+    $ docker service update mysql \
+                     --secret-add source=mysql_password_v2,target=mysql_password \
+                     --secret-rm mysql_password
+    ```
+
+    Move on to the next step quickly, because WordPress clients will not be able
+    to connect to the MySQL database until you update its password.
+
+3. Update the `wordpress` service to use the new password, keeping the target
+   path at `/run/secrets/wp_db_secret` and keeping the file permissions at `0400`.
+
+   ```bash
+   $ docker service update wordpress \
+                    --secret-add source=mysql_password_v2,target=wp_db_password,mode=0400 \
+                    --secret-rm mysql_password
+   ```
+
+4.  Verify that WordPress works by going to `http://localhost:30000/` again.
```

Comment:
Hm, okay, bad news; this example shows how to replace a secret, but unfortunately it doesn't work;

- both the `wordpress` and `mysql` services use an anonymous volume; each time these services are updated, a new volume is created, so any changes made are lost. After updating, I end up with a _fresh_ database, and a _fresh_ WordPress installation (i.e., uploads, and customizations are no longer there)
- the `mysql` image does not currently have an option to _update_ an existing password through environment variables. The provided secret/password is only used during first initialization, but ignored after that


To see that it doesn't work;

Create volumes in advance for both services (so that they keep modifications);

    docker volume create wp_data
    docker volume create mysql_data


Or implicitly create the volumes during `docker service create`;

    --mount source=mysql_data,target=/var/lib/mysql
    --mount source=wordpress_data,target=/var/www/html

```bash
docker service create --name="mysql" \
                        --mount source=mysql_data,target=/var/lib/mysql \
                        --network="wp_network" \
                        --replicas=1 \
                        --secret=mysql_password \
                        -e MYSQL_ROOT_PASSWORD_FILE=/run/secrets/mysql_password \
                        mysql

docker service create --name="wordpress" \
                        --mount source=wordpress_data,target=/var/www/html \
                        --network="wp_network" \
                        --replicas="1" \
                        --publish 30000:80 \
                        --secret source=mysql_password,target=wp_db_password,mode=0400 \
                        -e WORDPRESS_DB_PASSWORD_FILE="/run/secrets/wp_db_password" \
                        -e WORDPRESS_DB_HOST="mysql:3306" \
                        wordpress:latest

```bash

Now rotate the secrets;

```bash
docker service update mysql \
                 --secret-add source=mysql_password_v2,target=mysql_password \
                 --secret-rm mysql_password

docker service update wordpress \
                --secret-add source=mysql_password_v2,target=wp_db_password,mode=0400 \
                --secret-rm mysql_password
```


After this, the mysql service is running, but still uses the old password,
the WordPress service is not running, because it cannot connect to the
database

```
MySQL Connection Error: (1045) Access denied for user 'root'@'10.0.0.3' (using password: YES)
Warning: mysqli::mysqli(): (HY000/1045): Access denied for user 'root'@'10.0.0.3' (using password: YES) in - on line 19
```